core/services/keeper/registry_synchronizer_core.go

```diff
@@ -0,0 +1,140 @@
+package keeper
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated"
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/keeper_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/core/services"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/log"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"gorm.io/gorm"
+)
+
+// MailRoom holds the log mailboxes for all the log types that keeper cares about
+type MailRoom struct {
+	mbUpkeepCanceled   *utils.Mailbox
+	mbSyncRegistry     *utils.Mailbox
+	mbUpkeepPerformed  *utils.Mailbox
+	mbUpkeepRegistered *utils.Mailbox
+}
+
+func NewRegistrySynchronizer(
+	job job.Job,
+	contract *keeper_registry_wrapper.KeeperRegistry,
+	db *gorm.DB,
+	headBroadcaster *services.HeadBroadcaster,
+	logBoradcaster log.Broadcaster,
+	syncInterval time.Duration,
+	minConfirmations uint64,
+) *RegistrySynchronizer {
+	mailRoom := MailRoom{
+		mbUpkeepCanceled:   utils.NewMailbox(50),
+		mbSyncRegistry:     utils.NewMailbox(1),
+		mbUpkeepPerformed:  utils.NewMailbox(1),
+		mbUpkeepRegistered: utils.NewMailbox(50),
+	}
+	return &RegistrySynchronizer{
+		chHeads:          make(chan models.Head, 1),
```

Comment:
the mailbox is pretty cumbersome because you have to type-check upon retrieval. For simpler use cases like this I think a good 'ol fashion channel works best.